docs/source/proposals/typing_recursion.rst

```diff
@@ -0,0 +1,129 @@
+========================
+NBEP ?: Typing Recursion
+========================
+
+:Author: Siu Kwan Lam
+:Date: Sept 2016
+:Status: Draft
+
+Introduction
+============
+
+This document proposes an enhancement to the type inference to
+support recursion without explicit annotating the function signature.
+As a result, the proposal enables numba to type-infer both self-recursive and
+mutual-recursive functions under some limitations.  In practice, these
+limitions can be easily overcome by specifying a compilation order.
+
+
+The Current State
+=================
+
+Recursion support in numba is currently limited to self-recursion with explicit
+type annotation for the function.  This limitation comes from the inability to
+determine the return type of a recursive call.  This is because the callee is
+either the current function (for self-recursion) or a parent function
+(mutual-recursion).  Its type inference has been suspended while waiting for
+the function-type of its callee.  This forms a cyclic dependency.
+Given a function ``foo()`` that calls ``bar()``, which in turns call ``foo()``::
+
+    def foo(x):
+        if x > 0:
+            return bar(x)
+        else:
+            return 1
+
+    def bar(x):
+        return foo(x - 1)
+
+
+The type inference of ``foo()`` depends on that of ``bar()``, which depends on
+``foo()``.  Therefore ``foo()`` depends on itself and the type inference cannot
+terminate.
+
+
+The Solution
+============
+
+The proposed solution has two components:
+
+1. Introduce a compile-time *callstack* that tracks the compiling functions.
+2. Allow partial type inference on functions by leveraging the return type
+   on non-recursive control-flow paths.
+
+The compile-time callstack stores typing information of functions being
+compiled.  Like an ordinary callstack, it pushes a new record every time a
+function is "called".  Since this occurs at compile-time, a "call" triggers
+a compilation of the callee.
+
+To detect recursion, the compile-time callstack is searched bottom-up
+(stack grows downward) for a record that matches the callee.
+The record contains a reference to the type inference state.
+With that, type inference can be resumed to determine the return type.
+
+Recall that the type inference cannot be resumed normally because of the cyclic
```

Comment:
I think either `the type inference algorithm/mechanism` or just plain `type inference` should be used throughout ?